exchanges/bittrex/bittrex_websocket.go

```diff
@@ -0,0 +1,627 @@
+	if err != nil {
+		return err
+	}
+
+	params := url.Values{}
+	params.Set("clientProtocol", "1.5")
+	params.Set("transport", "webSockets")
+	params.Set("connectionToken", wsHandshakeData.ConnectionToken)
+	params.Set("connectionData", "[{name:\"c3\"}]")
+	params.Set("tid", "10")
+
+	path := common.EncodeURLValues("/connect", params)
+
+	err = b.Websocket.SetWebsocketURL(endpoint+path, false, false)
+	if err != nil {
+		return err
+	}
+
+	err = b.Websocket.Conn.Dial(&dialer, http.Header{})
+	if err != nil {
+		return err
+	}
+	// Can set up custom ping handler per websocket connection.
+	b.Websocket.Conn.SetupPingHandler(stream.PingHandler{
+		MessageType: websocket.PingMessage,
+		Delay:       bittrexWebsocketTimer,
+	})
+	err = b.seedOrderBook()
+	if err != nil {
+		return err
+	}
+	// This reader routine is called prior to initiating a subscription for
+	// efficient processing.
+	go b.wsReadData()
+	if b.GetAuthenticatedAPISupport(exchange.WebsocketAuthentication) {
+		err = b.WsAuth()
+		if err != nil {
+			b.Websocket.DataHandler <- err
+			b.Websocket.SetCanUseAuthenticatedEndpoints(false)
+		}
+	}
+	return nil
+}
+
+// WsSignalRHandshake requests the SignalR connection token over https
+func (b *Bittrex) WsSignalRHandshake(result interface{}) error {
+	endpoint, err := b.API.Endpoints.GetURL(exchange.WebsocketSpotSupplementary)
+	if err != nil {
+		return err
+	}
+	path := "/negotiate?connectionData=[{name:\"c3\"}]&clientProtocol=1.5"
+	return b.SendPayload(context.Background(), &request.Item{
+		Method:        http.MethodGet,
+		Path:          endpoint + path,
+		Result:        result,
+		Verbose:       b.Verbose,
+		HTTPDebugging: b.HTTPDebugging,
+		HTTPRecording: b.HTTPRecording,
+	})
+}
+
+// WsAuth sends an authentication message to receive auth data
+// Authentications expire after 10 minutes
+func (b *Bittrex) WsAuth() error {
+	// [apiKey, timestamp in ms, random uuid, signed payload]
+	apiKey := b.API.Credentials.Key
+	randomContent, err := uuid.NewV4()
+	if err != nil {
+		return err
+	}
+	timestamp := strconv.FormatInt(time.Now().UnixNano()/1000000, 10)
+	hmac := crypto.GetHMAC(
+		crypto.HashSHA512,
+		[]byte(timestamp+randomContent.String()),
+		[]byte(b.API.Credentials.Secret),
+	)
+	signature := crypto.HexEncodeToString(hmac)
+
+	invocationIDCounter++
+	request := WsEventRequest{
+		Hub:          "c3",
+		Method:       "Authenticate",
+		InvocationID: invocationIDCounter,
+	}
+
+	arguments := make([]string, 0)
+	arguments = append(arguments, apiKey, timestamp, randomContent.String(), signature)
+	request.Arguments = arguments
+
+	requestString, err := json.Marshal(request)
+	if err != nil {
+		return err
+	}
+	if b.Verbose {
+		log.Debugf(log.ExchangeSys, "%s Sending JSON message - %s\n", b.Name, requestString)
+	}
+	err = b.Websocket.Conn.SendJSONMessage(request)
+	if err != nil {
+		return err
+	}
+
+	b.WsPendingRequests[request.InvocationID] = WsPendingRequest{
+		request,
+		nil,
```

Comment:
remove please.

Reply:
Remove what exactly? And why?

Reply:
Just the line, we dont need to see the value as nil.

Reply:
Ah. It actually does want a value; without the `nil`:
```
# github.com/thrasher-corp/gocryptotrader/exchanges/bittrex
exchanges/bittrex/bittrex_websocket.go:176:63: too few values in WsPendingRequest literal
```

Reply:
```go
b.WsPendingRequests[request.InvocationID] = WsPendingRequest{
	WsEventRequest: request,
}
```

Reply:
I'm resolving this as it's really petty. 😆

Reply:
Oh and now I understand ;)